cli/cmd/top.go

```diff
@@ -0,0 +1,393 @@
+  linkerd top pod/web-dlbvj`,
+		Args:      cobra.RangeArgs(1, 2),
+		ValidArgs: util.ValidTargets,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			requestParams := util.TapRequestParams{
+				Resource:    strings.Join(args, "/"),
+				Namespace:   options.namespace,
+				ToResource:  options.toResource,
+				ToNamespace: options.toNamespace,
+				MaxRps:      options.maxRps,
+				Scheme:      options.scheme,
+				Method:      options.method,
+				Authority:   options.authority,
+				Path:        options.path,
+			}
+
+			req, err := util.BuildTapByResourceRequest(requestParams)
+			if err != nil {
+				return err
+			}
+
+			return getTrafficByResourceFromAPI(os.Stdout, validatedPublicAPIClient(), req)
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace,
+		"Namespace of the specified resource")
+	cmd.PersistentFlags().StringVar(&options.toResource, "to", options.toResource,
+		"Display requests to this resource")
+	cmd.PersistentFlags().StringVar(&options.toNamespace, "to-namespace", options.toNamespace,
+		"Sets the namespace used to lookup the \"--to\" resource; by default the current \"--namespace\" is used")
+	cmd.PersistentFlags().Float32Var(&options.maxRps, "max-rps", options.maxRps,
+		"Maximum requests per second to tap.")
+	cmd.PersistentFlags().StringVar(&options.scheme, "scheme", options.scheme,
+		"Display requests with this scheme")
+	cmd.PersistentFlags().StringVar(&options.method, "method", options.method,
+		"Display requests with this HTTP method")
+	cmd.PersistentFlags().StringVar(&options.authority, "authority", options.authority,
+		"Display requests with this :authority")
+	cmd.PersistentFlags().StringVar(&options.path, "path", options.path,
+		"Display requests with paths that start with this prefix")
+
+	return cmd
+}
+
+func getTrafficByResourceFromAPI(w io.Writer, client pb.ApiClient, req *pb.TapByResourceRequest) error {
+
+	rsp, err := client.TapByResource(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	err = termbox.Init()
+	if err != nil {
+		return err
+	}
+	defer termbox.Close()
+
+	requestCh := make(chan topRequest, 100)
+	done := make(chan struct{})
+
+	go recvEvents(rsp, requestCh, done)
+	go pollInput(done)
+
+	renderTable(requestCh, done)
+
+	return nil
+}
+
+func recvEvents(tapClient pb.Api_TapByResourceClient, requestCh chan<- topRequest, done chan<- struct{}) {
+	outstandingRequests := make(map[topRequestID]topRequest)
+	for {
+		event, err := tapClient.Recv()
+		if err == io.EOF {
+			fmt.Println("Tap stream terminated")
+			close(done)
+			return
+		}
+		if err != nil {
+			fmt.Println(err.Error())
+			close(done)
+			return
+		}
+
+		switch ev := event.GetHttp().GetEvent().(type) {
+		case *pb.TapEvent_Http_RequestInit_:
+			id := topRequestID{
+				addr.PublicAddressToString(event.GetSource()),
+				addr.PublicAddressToString(event.GetDestination()),
+				ev.RequestInit.GetId().Stream,
+			}
```

Comment:
One more small go style nit -- for multi-line constructors, we tend to use named fields so that there is less chance of field misalignment. So I'd write this as:

```
			id := topRequestID{
				src:    addr.PublicAddressToString(event.GetSource()),
				dst:    addr.PublicAddressToString(event.GetDestination()),
				stream: ev.RequestInit.GetId().Stream,
			}
```

Furthermore, you're initializing the ID in each branch of the case statement, but the `src` and `dst` fields don't change. So you could save yourself a few lines by doing something like:

```
		id := topRequestID{
			src: addr.PublicAddressToString(event.GetSource()),
			dst: addr.PublicAddressToString(event.GetDestination()),
		}

		switch ev := event.GetHttp().GetEvent().(type) {
		case *pb.TapEvent_Http_RequestInit_:
			id.stream = ev.RequestInit.GetId().Stream
			outstandingRequests[id] = topRequest{
				event:   event,
				reqInit: ev.RequestInit,
			}

		case *pb.TapEvent_Http_ResponseInit_:
			id.stream = ev.ResponseInit.GetId().Stream
			if req, ok := outstandingRequests[id]; ok {
				req.rspInit = ev.ResponseInit
				outstandingRequests[id] = req
			} else {
				log.Warnf("Got ResponseInit for unknown stream: %s", id)
			}

		case *pb.TapEvent_Http_ResponseEnd_:
			id.stream = ev.ResponseEnd.GetId().Stream
			if req, ok := outstandingRequests[id]; ok {
				req.rspEnd = ev.ResponseEnd
				requestCh <- req
			} else {
				log.Warnf("Got ResponseEnd for unknown stream: %s)", id)
			}
		}
```

Since not all fields need to be initialized when the struct is instantiated.